### PR TITLE
Feature nondet array initialization rebase

### PR DESCRIPTION
**Pull request by hannes-steffenhagen-diffblue**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

Adds nondet initialisation for statically sized arrays anywhere, and dynamically sized arrays as function arguments when explicitly requested. Dynamically sized arrays can also have an associated size parameter that will be set to the size of the dynamic array.

This is based on a (rebased) version of https://github.com/diffblue/cbmc/pull/3251, new commits start at 0a590c. Code should be reviewable as is, however commits still need some cleanup and there'll need to be some additional documentation before it's mergeable, hence the WIP tag.

### CODE REVIEW

**Comment by hannes-steffenhagen-diffblue** [comment]
Heads up, tests failures seem to be because we apparently expand `NULL` to `(void*)0` now _before_ generating the assert string. Also, I'll address formatting complaints from cpplint/clang-format when I get around to fixing up the commits

**Review by xbauch — APPROVED** [review]

File: src/cbmc/cbmc_parse_options.cpp [hunk 1]
```diff
@@ -928,6 +929,20 @@ void cbmc_parse_optionst::help()
     " --round-to-plus-inf          rounding towards plus infinity\n"
     " --round-to-minus-inf         rounding towards minus infinity\n"
     " --round-to-zero              rounding towards zero\n"
+    " --max-nondet-tree-depth N    limit size of nondet (e.g. input) object tree;\n" /* NOLINT(*) */ \
+    "                              at level N pointers are set to null\n" /* NOLINT(*) */ \
+    " --min-null-tree-depth N      minimum level at which a pointer can first be\n" /* NOLINT(*) */ \
+    " --pointers-to-treat-as-array <identifier,...>  Comma separated list of identifiers that should be initialized" \
```
Comment (danpoe) [inline src/cbmc/cbmc_parse_options.cpp]: Missing newline

File: src/cbmc/cbmc_parse_options.cpp [hunk 2]
```diff
@@ -928,6 +929,20 @@ void cbmc_parse_optionst::help()
...
+    "                               as arrays" \
+    "                              NULL in a recursively nondet initialized struct\n" /* NOLINT(*) */ \
+    " --associated-array-sizes <identifier:identifier...>"
```
Comment (danpoe) [inline src/cbmc/cbmc_parse_options.cpp]: Missing newline

File: regression/cbmc/pointer-to-array-function-parameters-multi-arg-right/test.c [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  const char *string,
+  int string_size,
+  const char *prefix,
+  int prefix_size)
+{
+  if(string_size < prefix_size)
+  {
+    return 0;
+  }
+  // oh no! we should have used prefix_size here
```
Comment (danpoe) [inline regression/cbmc/pointer-to-array-function-parameters-multi-arg-right/test.c]: Should be removed

File: src/ansi-c/c_object_factory_parameters.h [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  {
+  }
+
+  bool should_be_treated_as_array(irep_idt id) const;
+  bool is_array_size_parameter(irep_idt id) const;
+  optionalt<irep_idt> get_associated_size_variable(irep_idt array_id) const;
+
+  void set(const optionst &options) override;
+
+  std::size_t max_dynamic_array_size = 10;
```
Comment (danpoe) [inline src/ansi-c/c_object_factory_parameters.h]: 10 seems already a bit high for a default value. Can we use 2?

File: src/ansi-c/c_object_factory_parameters.h [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+Author: Daniel Poetzl
+
+\*******************************************************************/
+
+#ifndef CPROVER_ANSI_C_C_OBJECT_FACTORY_PARAMETERS_H
+#define CPROVER_ANSI_C_C_OBJECT_FACTORY_PARAMETERS_H
+
+#include <algorithm>
+#include <string>
```
Comment (danpoe) [inline src/ansi-c/c_object_factory_parameters.h]: Can we move `<algorithm>` and `<string>` to the cpp file?

File: src/ansi-c/c_object_factory_parameters.h [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  bool is_array_size_parameter(irep_idt id) const;
+  optionalt<irep_idt> get_associated_size_variable(irep_idt array_id) const;
+
+  void set(const optionst &options) override;
+
+  std::size_t max_dynamic_array_size = 10;
+
+private:
+  std::vector<irep_idt> pointers_to_treat_as_array;
+  std::set<irep_idt> variables_that_hold_array_sizes;
```
Comment (danpoe) [inline src/ansi-c/c_object_factory_parameters.h:37]: Could `pointers_to_treat_as_array` be a set as well?

File: src/ansi-c/c_object_factory_parameters.cpp [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+  {
+    array_name_to_associated_array_size_variable.clear();
+    variables_that_hold_array_sizes.clear();
+    auto const &array_size_pairs =
+      options.get_list_option("associated-array-sizes");
+    for(auto const &array_size_pair : array_size_pairs)
+    {
+      std::string array_name;
+      std::string size_name;
+      split_string(array_size_pair, ':', array_name, size_name);
```
Comment (danpoe) [inline src/ansi-c/c_object_factory_parameters.cpp]: We should probably catch the `deserialization_exceptiont` here and then throw an `invalid_user_input_exceptiont`.

File: src/util/optional_utilities.h [hunk 8]
```diff
@@ -0,0 +1,23 @@
+/*******************************************************************\
```
Comment (danpoe) [inline src/util/optional_utilities.h]: Nit: these files are typically called *_utils (not *_utilities).

File: src/ansi-c/c_nondet_symbol_factory.cpp [hunk 9]
```diff
@@ -127,18 +133,53 @@ exprt symbol_factoryt::allocate_object(
...
+      {
+        gen_nondet_size_array_init(
+          assignments, symbol_expr, depth, recursion_set);
+        return;
+      } else if(object_factory_params.is_array_size_parameter(symbol_name))
+      {
+        // skip, we'll handle this during array initialisation
+        return;
+      }
+    }
```
Comment (danpoe) [inline src/ansi-c/c_nondet_symbol_factory.cpp:193]: Array initialization is done here only if `type.id() == ID_pointer` and `expr.id() == ID_symbol`. However, when recursively initializing a struct a member expression is passed to `gen_nondet_init()`. Does this mean that arrays are only nondet initialized when they are function parameters (but not when they are within recursively initialized structs)?

File: src/util/string_utils.cpp [hunk 10]
```diff
@@ -109,7 +110,14 @@ void split_string(
...
 
   split_string(s, delim, result, strip);
-  CHECK_RETURN(result.size() == 2);
+  if(result.size() != 2)
+  {
+    throw deserialization_exceptiont{"expected string `" + s +
+                                     "' to contain two substrings "
+                                     "delimited by " +
+                                     delim + " but has " +
+                                     std::to_string(result.size())};
```
Comment (danpoe) [inline src/util/string_utils.cpp:119]: The end of the error message is a bit unclear. Maybe change to something like "... but have \<n\> substrings".

**Review by danpoe — COMMENTED** [review]
Looks good so far, a few minor comments

File: regression/cbmc/pointer-to-array-function-parameters-multi-arg-right/test.c (hunk 3 above)
Comment (hannes-steffenhagen-diffblue) [reply regression/cbmc/pointer-to-array-function-parameters-multi-arg-right/test.c]: Ah sorry, this was meant to be documentation for what's "wrong" with this test case. Maybe too tongue-in-cheek, I'll reword it

File: src/ansi-c/c_object_factory_parameters.h (hunk 4 above)
Comment (hannes-steffenhagen-diffblue) [reply src/ansi-c/c_object_factory_parameters.h]: We can

File: regression/cbmc/pointer-to-array-function-parameters-multi-arg-right/test.c (hunk 3 above)
Comment (danpoe) [reply regression/cbmc/pointer-to-array-function-parameters-multi-arg-right/test.c]: I think the wording is fine actually, but it seems the comment is wrong for this file (looks like it's been copied from another file).

File: src/ansi-c/c_nondet_symbol_factory.cpp (hunk 9 above)
Comment (hannes-steffenhagen-diffblue) [reply src/ansi-c/c_nondet_symbol_factory.cpp:193]: Right, this is a planned feature, but I've been told to put that on the backburner for now.

For the record, I was thinking of handling that with having a syntax like `struct_name::member_name` for `--pointers-to-treat-as-arrays` and `--associated-array-sizes` (I'd only allow sizes and pointers that are in the same struct for this). That'd presumably have to use struct tags though, which might be a bit annoying for users.

**Comment by hannes-steffenhagen-diffblue** [comment]
Also, turns out test failures where also because SMT2 solver is broken. I'll rebase, reorganise and squash some of the commits in here now, so hopefully it'll be review-ready then

File: regression/cbmc/pointer-to-array-function-parameters-multi-arg-right/test.c (hunk 3 above)
Comment (hannes-steffenhagen-diffblue) [reply regression/cbmc/pointer-to-array-function-parameters-multi-arg-right/test.c]: Oh. You're right, sorry didn't see that!

**Pull request closed without merging**